src/decoder/mod.rs

```diff
@@ -424,11 +424,13 @@ impl<R: Read + Seek> Decoder<R> {
     /// Reads a string
     #[inline]
     pub fn read_string(&mut self, length: usize) -> TiffResult<String> {
-        let mut out = String::with_capacity(length);
-        self.reader.read_to_string(&mut out)?;
+        let mut out = vec![0; length];
+        self.reader.read_exact(&mut out)?;
         // Strings may be null-terminated, so we trim anything downstream of the null byte
-        let trimmed = out.bytes().take_while(|&n| n != 0).collect::<Vec<u8>>();
-        Ok(String::from_utf8(trimmed)?)
+        if let Some(first) = out.iter().position(|&b| b == 0) {
+            out.truncate(first);
+        }
+        Ok(String::from_utf8(out)?)
```

Comment:
This change is because otherwise we'd require utf-8 encoding after the null terminator, right?
Do you have a small regression test? Would it be possible to create one?

Reply:
It happen when i encode -9999 in the ASCII tag with gdal: `gdal_edit.py -a_nodata -9999 resources/zh_dem_25.tif`

The output bytes are `[0, 0, 0, 0, 0, 0]`, so i think creating a tag with the same value should create the bug we had with invalid utf-8 string.

Reply:
The byte `\0` is a valid utf-8 character though, so that should work with the new and old code alike.

Reply:
Yeah, i can keep the same code and just change the out into iter:
```rust
// Strings may be null-terminated, so we trim anything downstream of the null byte
let trimmed = out.into_iter().take_while(|&n| n != 0).collect::<Vec<u8>>();
```
The main issue was the `read_to_string` function.

Reply:
I don't understand? `truncate` is surely more performant than collecting individual bytes in a new vector. No, this current code is definitely more correct. I was only unsure how this issue was found in the first case since `read_to_string` should have no problem with reading `[0, 0, 0, 0, 0, 0]`.

Reply:
Excuse me the value is `[45, 57, 57, 57, 57, 0]`
Ok i thought your impl was more performant than mine.

Reply:
Are there by any chance any bytes after the null terminator?

Reply:
I'm still not sure why this would fail with the previous implementation otherwise.

Reply:
I also don't understand why this fail with the method read_to_string
Here is more bytes:
`[45, 57, 57, 57, 57, 0, 0, 0, 0, 0, 0, 0, 57, 64, 0, 0, 0, 0, 0, 0, 57, 64, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]`

Reply:
Yeah, if any of the values were larger than `128` then it might no longer be valid utf8. Such a tail would be rejected by `read_to_string` but accepted by the new variant, which seems more correct. In any case I was only trying to understand if it were easy to create a reduced regression test (<4kB) for this. It's not critical for the PR.